components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServiceURL.java

```diff
@@ -0,0 +1,252 @@
+     *
+     * @return String of the url path.
+     */
+    public String getUrlPath() {
+
+        return urlPath;
+    }
+
+    /**
+     * This method is called to get the parameter value when the key is provided.
+     *
+     * @param key Key of the parameter.
+     * @return The value of the parameter.
+     */
+    public String getParameter(String key) {
+
+        return parameters.get(key);
+    }
+
+    /**
+     * This method is called to get a list of the parameter names.
+     *
+     * @return ArrayList of parameter keys.
+     */
+    public String[] getParameterKeys() {
+
+        ArrayList<String> parameterArrayList = new ArrayList<>();
+
+        for (Map.Entry<String, String> entry : this.parameters.entrySet()) {
+            parameterArrayList.add(entry.getKey());
+        }
+        return parameterArrayList.toArray(new String[0]);
+    }
+
+    /**
+     * This method is called to get the decoded fragment from the url.
+     *
+     * @return The decoded fragment.
+     * @throws URLBuilderException
+     */
+    public String getFragment() {
+
+        return fragment;
+    }
+
+    /**
+     * Concatenate the protocol, host name, port, proxy context path, web context root and url context to return the
+     * absolute URL.
+     *
+     * @return The absolute URL from the {@link ServiceURL} instance.
+     */
+    public String getAbsoluteURL() {
+
+        return this.absoluteUrl;
+    }
+
+    /**
+     * This method is called to get the relative url from the url context.
+     *
+     * @return The relative URL from the {@link ServiceURL} instance.
+     */
+    public String getRelativeURL() {
+
+        return this.relativeUrl;
+    }
+
+    private void appendContextToUri(StringBuilder serverUrl, String contextPath) {
+
+        if (StringUtils.isNotBlank(contextPath)) {
+            if (contextPath.endsWith("/")) {
+                contextPath = contextPath.substring(0, contextPath.length() - 1);
+            }
+            if (StringUtils.isNotBlank(contextPath) && contextPath.trim().charAt(0) != '/') {
+                serverUrl.append("/").append(contextPath.trim());
+            } else {
+                serverUrl.append(contextPath.trim());
+            }
+        }
+    }
+
+    private void appendParamsToUri(StringBuilder serverUrl, String resolvedParamsString, String delimiter)
+            throws URLBuilderException {
+
+        if (serverUrl.length() > 0 && serverUrl.charAt(serverUrl.length() - 1) == '/') {
+            serverUrl.setLength(serverUrl.length() - 1);
+        }
+        if (StringUtils.isNotBlank(resolvedParamsString)) {
+            try {
+                serverUrl.append(delimiter).append(URLEncoder.encode(resolvedParamsString,
+                        StandardCharsets.UTF_8.name()));
+            } catch (UnsupportedEncodingException e) {
+                throw new URLBuilderException("Error while trying to build the url", e);
```

Comment:
Add more context info here

Reply:
Shouldn't we return with an error code. @mefarazath what was the approach followed previously. Was it in API level we returned with error codes